### PR TITLE
connectedhomeip: add door-lock sample

### PR DESCRIPTION
**Pull request by Damian-Nordic**
Add lock sample similar to lock-app from the CHIP repository
with a few changes:
1. remove partially implemented features like auto re-lock
   or software update which increase code complexity and
   don't bring much value at the moment
2. refactor event dispatching to make it easier to read
3. conform to NCS/Zephyr code style guidelines.

A device running the application works as a CHIP accessory.
That is, it can be commissioned over Bluetooth LE onto
a Thread network, and then controlled using CHIP messages
sent over the Thread network.

Note that code in server/ will eventually be moved to the CHIP library!

Signed-off-by: Damian Krolik <damian.krolik@nordicsemi.no>

### CODE REVIEW

File: samples/connectedhomeip/lock/README.rst
Comment (kkasperczyk-no) [inline samples/connectedhomeip/lock/README.rst:11]: Links should not be used directly, but added to the ncs/nrf/doc/nrf/links.txt and referenced to its name.

File: samples/connectedhomeip/lock/README.rst
Comment (kkasperczyk-no) [inline samples/connectedhomeip/lock/README.rst:10]: One sentence per line, without breaking it, as @greg-fer said.

File: samples/connectedhomeip/lock/src/server/RendezvousServer.cpp
Comment (kkasperczyk-no) [inline samples/connectedhomeip/lock/src/server/RendezvousServer.cpp:7]: I guess it's temporary solution and files from src/server will be removed after merging chip-app-server integration PR?

File: samples/connectedhomeip/lock/src/util/ThreadUtil.h
Comment (kkasperczyk-no) [inline samples/connectedhomeip/lock/src/util/ThreadUtil.h:7]: I'm not sure about using `#pragma once` in sdk-nrf. I'm not saying it's wrong, but I think that it is not used in any other place, so some assumptions on this could be made.

File: samples/connectedhomeip/lock/src/util/LEDWidget.cpp
Comment (kkasperczyk-no) [inline samples/connectedhomeip/lock/src/util/LEDWidget.cpp:6]: Do you think it would be good idea to move src/util content to some common directory for all samples, as soon we will create another one using the same files?

**Review by kkasperczyk-no — COMMENTED** [review]
I added some partial review comments.

**Comment by kkasperczyk-no** [comment]
@greg-fer FYI

File: samples/connectedhomeip/lock/src/server/RendezvousServer.cpp
Comment (Damian-Nordic) [reply samples/connectedhomeip/lock/src/server/RendezvousServer.cpp:7]: Yes, I even mentioned that in the PR description ;)

File: samples/connectedhomeip/lock/src/util/LEDWidget.cpp
Comment (Damian-Nordic) [reply samples/connectedhomeip/lock/src/util/LEDWidget.cpp:6]: Yes, I can do that, I guess.

File: samples/connectedhomeip/lock/src/util/ThreadUtil.h
Comment (Damian-Nordic) [reply samples/connectedhomeip/lock/src/util/ThreadUtil.h:7]: I'm not sure either. Can't find any hints on that in Zephyr/Linux Coding Style guides which NCS coding style is based on. I guess some NCS code owners will decide when we push the changes to main repo.

File: samples/connectedhomeip/lock/README.rst
Comment (Damian-Nordic) [reply samples/connectedhomeip/lock/README.rst]: Done

File: samples/connectedhomeip/lock/README.rst
Comment (Damian-Nordic) [reply samples/connectedhomeip/lock/README.rst]: Done

File: samples/connectedhomeip/lock/src/util/LEDWidget.cpp
Comment (Damian-Nordic) [reply samples/connectedhomeip/lock/src/util/LEDWidget.cpp]: Done

File: samples/connectedhomeip/lock/README.rst
Comment (kkasperczyk-no) [reply samples/connectedhomeip/lock/README.rst]: I meant replacing all links in file, like these in the "Commissioning and remote control" section. Sorry that I didn't write it clearly.

**Review by kkasperczyk-no — APPROVED** [review]
It looks good, I don't have any comments on implementation. Please fix all of the links in the docs and that's all from me.

File: samples/connectedhomeip/lock/README.rst
Comment (greg-fer) [inline samples/connectedhomeip/lock/README.rst]: @Damian-Nordic , could you add some output lines to these steps, as discussed f2f? Thanks!

**Review by greg-fer — APPROVED** [review]
Please check my commit, squash it, and add sample output.

**Review by greg-fer — APPROVED** [review]

**Pull request merged**
